lms/extractors/notebook.py

```diff
@@ -1,46 +1,65 @@
-from itertools import chain
+import itertools
 import json
-from operator import itemgetter
 from typing import Any, Dict, Iterator, List, Tuple
 
-from lms.extractors.base import Extractor
+from lms.extractors.base import Extractor, File
 
 
 NotebookJson = Dict[str, Any]
 Cell = Dict[str, Any]
 
 
 class Notebook(Extractor):
+    POSSIBLE_JSON_EXCEPTIONS = (
+        json.JSONDecodeError, KeyError, StopIteration, UnicodeDecodeError,
+    )
+
     def __init__(self, **kwargs):
         super().__init__(**kwargs)
         try:
             cells = self._get_code_cells()
-            self.cells = chain([next(cells)], cells)  # Run the generator
-        except (json.JSONDecodeError, KeyError):
+            # Triggers StopIteration if `cells` is empty (see example below).
+            first_cell = next(cells)
+            self.cells = itertools.chain([first_cell], cells)
+        except self.POSSIBLE_JSON_EXCEPTIONS:
             self.is_json = False
         else:
             self.is_json = True
 
     def can_extract(self) -> bool:
         return self.is_json
 
-    def _get_code_cells(self) -> Iterator[Cell]:
-        notebook = json.loads(self.to_extract)
-        cells = notebook['cells']
-        yield from filter(self._is_code_cell, cells)
-
     @staticmethod
     def _is_code_cell(cell: Cell) -> bool:
         return (
             cell.get('cell_type', '') == 'code'
             and bool(cell.get('source'))
         )
 
-    @classmethod
-    def get_exercise(cls, to_extract: Cell) -> Tuple[str, str]:
+    def _get_code_cells(self) -> Iterator[Cell]:
+        notebook = json.loads(self.file_content)
+        cells = notebook['cells']
+        yield from filter(self._is_code_cell, cells)
+
+    def get_exercise(self, to_extract: Cell) -> Tuple[int, List[File]]:
         code: List[str] = to_extract.get('source', [])
-        return cls._clean(code)
+        exercise_id, clean_code = self._clean(code)
```

Comment:
you don't want to check if `clean_code` is not empty or `exercise_id` is not `0`? you may get empty files this way

Reply:
Nice catch!